File: sample-conference-videochat/sample-conference-videochat/Utils/Categories/UIImage+Cropper.m
//
//  UIImage+Cropper.m
//  sample-conference-videochat
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "UIImage+Cropper.h"

@implementation UIImage (Cropper)

- (UIImage *)imageWithCornerRadius:(CGFloat)cornerRadius
                        targetSize:(CGSize)targetSize {
    
    UIImage *scaledImage = [self imageByScaleAndCrop:targetSize];
    
    float scaleFactor = [[UIScreen mainScreen] scale];
    UIGraphicsBeginImageContextWithOptions(scaledImage.size, NO, scaleFactor);

    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextTranslateCTM(context, 0, scaledImage.size.height);
    CGContextScaleCTM(context, 1.0, -1.0);

    UIBezierPath *path = [UIBezierPath bezierPathWithRoundedRect:CGRectMake(0, 0, scaledImage.size.width, scaledImage.size.height)
                                               byRoundingCorners:UIRectCornerAllCorners
                                                     cornerRadii:CGSizeMake(cornerRadius, cornerRadius)];
    
    CGContextAddPath(context, path.CGPath);
    CGContextClip(context);
    
    CGContextDrawImage(context, CGRectMake(0, 0, scaledImage.size.width, scaledImage.size.height), scaledImage.CGImage);

    UIImage *roundedImage = UIGraphicsGetImageFromCurrentImageContext();
    
    UIGraphicsEndImageContext();
    
    return roundedImage;
}

- (UIImage *)imageByScaleAndCrop:(CGSize)targetSize {
    
    UIImage *newImage = nil;
    
    CGFloat targetWidth = targetSize.width;
    CGFloat targetHeight = targetSize.height;
    
    CGFloat scaleFactor = 0.0;
    CGFloat scaledWidth = targetWidth;
    CGFloat scaledHeight = targetHeight;
    
    CGPoint thumbnailPoint = CGPointMake(0.0, 0.0);
    
    if (!CGSizeEqualToSize(self.size, targetSize)) {
        
        CGFloat widthFactor = targetWidth / self.size.width;
        CGFloat heightFactor = targetHeight / self.size.height;
        
        if (widthFactor > heightFactor) {
            scaleFactor = widthFactor;
        }
        else {
            
            scaleFactor = heightFactor;
        }
        
        scaledWidth  = self.size.width * scaleFactor;
        scaledHeight = self.size.height * scaleFactor;

        if (widthFactor > heightFactor) {
            
            thumbnailPoint.y = (targetHeight - scaledHeight) * 0.5;
            
        } else {
            
            thumbnailPoint.x = (targetWidth - scaledWidth) * 0.5;
        }
    }
    UIGraphicsBeginImageContextWithOptions(targetSize, NO, 0.0);
    
    CGRect thumbnailRect = CGRectZero;
    thumbnailRect.origin = thumbnailPoint;
    thumbnailRect.size.width  = scaledWidth;
    thumbnailRect.size.height = scaledHeight;
    
    [self drawInRect:thumbnailRect];
    
    newImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return newImage ;
}

- (UIImage *)imageByCircularScaleAndCrop:(CGSize)targetSize {
    float scaleFactor = [[UIScreen mainScreen] scale];
    
    if (CGSizeEqualToSize(targetSize, CGSizeZero)) {
        targetSize = self.size;
    }
    
    CGColorSpaceRef colorSpace =
    CGColorSpaceCreateDeviceRGB();
    CGContextRef context =
    CGBitmapContextCreate(NULL,
                          targetSize.width * scaleFactor,
                          targetSize.height * scaleFactor,
                          8,
                          targetSize.width * scaleFactor * 4,
                          colorSpace,
                          kCGImageAlphaPremultipliedFirst);
    
    CGContextScaleCTM(context,
                      scaleFactor,
                      scaleFactor);
    
    CGContextBeginPath(context);
    
    CGContextAddArc(context,
                    targetSize.width / 2,
                    targetSize.height / 2,
                    targetSize.width / 2,
                    0,
                    2 * M_PI,
                    0);
    
    CGContextClosePath(context);
    
    CGFloat widthFactor = targetSize.width / self.size.width;
    CGFloat heightFactor = targetSize.height  / self.size.height;
    
    if (widthFactor > heightFactor) {
        scaleFactor = widthFactor;
    }
    else {
        
        scaleFactor = heightFactor;
    }
    
    float w = self.size.width * scaleFactor;
    float h = self.size.height * scaleFactor;
    
    CGContextClip(context);
    CGContextDrawImage(context,
                       CGRectMake(0,0, w, h),
                       self.CGImage);
    
    CGImageRef renderedImage =
    CGBitmapContextCreateImage(context);
    CGColorSpaceRelease(colorSpace);
    CGContextRelease(context);
    
    UIImage *image =
    [UIImage imageWithCGImage:renderedImage
                        scale:0
                  orientation:self.imageOrientation];
    CGImageRelease(renderedImage);
    
    return image;
}

@end

File: Swift-QMChatViewController/ChatViewController/ChatViewController/Extentions/UIImage+Cropper.swift
//
//  UIImage+Cropper.swift
//  Swift-QMChatViewController
//
//  Created by Vladimir Nybozhinsky on 11/15/18.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit

extension UIImage {
  
  func withCornerRadius(_ cornerRadius: CGFloat, targetSize: CGSize) -> UIImage? {
    
    guard let scaledImage = byScaleAndCrop(targetSize), let scaledImageCgImage = scaledImage.cgImage,
      let context = UIGraphicsGetCurrentContext()  else {
      return UIImage()
    }
    
    let scaleFactor = Float(UIScreen.main.scale)
    UIGraphicsBeginImageContextWithOptions(scaledImage.size, false, CGFloat(scaleFactor))
    
    // Build a context that's the same dimensions as the new size
    context.translateBy(x: 0.0, y: scaledImage.size.height)
    context.scaleBy(x: 1.0, y: -1.0)
    
    // Create a clipping path with rounded corners
    let path = UIBezierPath(roundedRect: CGRect(x: 0.0, y: 0.0, width: scaledImage.size.width,
                                                height: scaledImage.size.height ),
                            byRoundingCorners: .allCorners,
                            cornerRadii: CGSize(width: cornerRadius, height: cornerRadius))
    
    context.addPath(path.cgPath)
    context.clip()
    
    let imageRect = CGRect(x: 0.0, y: 0.0, width: scaledImage.size.width, height: scaledImage.size.height)
    context.draw(scaledImageCgImage, in: imageRect)
    
    // Draw the image to the context; the clipping path will make anything outside the rounded rect transparent
    
    // Create a CGImage from the context
    let roundedImage = UIGraphicsGetImageFromCurrentImageContext()
    
    UIGraphicsEndImageContext()
    
    return roundedImage
  }
  
  func image(byScaleAndCrop targetSize: CGSize) -> UIImage? {

    let targetWidth: CGFloat = targetSize.width
    let targetHeight: CGFloat = targetSize.height
    
    var scaleFactor: CGFloat = 0.0
    var scaledWidth: CGFloat = targetWidth
    var scaledHeight: CGFloat = targetHeight
    
    var thumbnailPoint = CGPoint(x: 0.0, y: 0.0)
    
    if size.equalTo(targetSize) == false {
      
      let widthFactor: CGFloat = targetWidth / size.width
      let heightFactor: CGFloat = targetHeight / size.height
      
      if widthFactor > heightFactor {
        scaleFactor = widthFactor
      } else {
        scaleFactor = heightFactor
      }
      
      scaledWidth = size.width * scaleFactor
      scaledHeight = size.height * scaleFactor
      
      // center the image
      
      if widthFactor > heightFactor {
        thumbnailPoint.y = (targetHeight - scaledHeight) * 0.5
      } else {
        thumbnailPoint.x = (targetWidth - scaledWidth) * 0.5
      }
    }
    // this is actually the interesting part:
    UIGraphicsBeginImageContextWithOptions(targetSize, false, 0.0)
    
    var thumbnailRect = CGRect.zero
    thumbnailRect.origin = thumbnailPoint
    thumbnailRect.size.width = scaledWidth
    thumbnailRect.size.height = scaledHeight
    
    draw(in: thumbnailRect)
    
    let newImage = UIGraphicsGetImageFromCurrentImageContext()
    UIGraphicsEndImageContext()
    
    return newImage
  }
  
  func image(byCircularScaleAndCrop targetSize: CGSize) -> UIImage? {
    var targetSize = targetSize
    //bitmap context properties
    var scaleFactor = CGFloat(UIScreen.main.scale)
    
    if targetSize.equalTo(CGSize.zero) {
      targetSize = size
    }
    
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    guard let context = CGContext(data: nil, width: Int(targetSize.width * CGFloat(scaleFactor)),
                                  height: Int(targetSize.height * CGFloat(scaleFactor)),
                                  bitsPerComponent: 8,
                                  bytesPerRow: Int(targetSize.width * CGFloat(scaleFactor) * 4),
                                  space: colorSpace,
                                  bitmapInfo: CGImageAlphaInfo.premultipliedFirst.rawValue),
      let cgImage = self.cgImage else {
                                    return UIImage()
    }
    
    context.scaleBy(x: scaleFactor, y: scaleFactor)
    
    context.beginPath()
    
    context.addArc(center: CGPoint(x: targetSize.width / 2, y: targetSize.height / 2),
                   radius: targetSize.width / 2, startAngle: 0, endAngle: 2 * .pi, clockwise: false)
    
    context.closePath()
    
    let widthFactor: CGFloat = targetSize.width / size.width
    let heightFactor: CGFloat = targetSize.height / size.height
    
    if widthFactor > heightFactor {
      scaleFactor = widthFactor
    } else {
      scaleFactor = heightFactor
    }
    
    let width: CGFloat = size.width * scaleFactor
    let height: CGFloat = size.height * scaleFactor
    
    context.clip()
    //draw image into bitmap context
    
    context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
    
    guard let renderedImage = context.makeImage() else {
      return UIImage()
    }
//    CGContextRelease(context)
    
    let image = UIImage(cgImage: renderedImage, scale: 0, orientation: imageOrientation)
//    CGImageRelease(renderedImage)
    
    return image
  }
}
